### ISSUE TITLE
CUDA not working for SAC

### ISSUE THREAD

**Issue by Zeus3101** [SforAiDl/genrl#162]

**Comment by Sharad24** [comment]
Is this still an issue?

**Comment by sampreet-arthi** [comment]
I'll resolve it while working on Off Policy Agents.

**Comment by IshanDaga** [comment]
For this, we just need to check for CUDA support on the system and then use the CUDA functions for computation right?

**Comment by Sharad24** [comment]
Yes. Just ensure things that can be on the GPU, are on the GPU. E.g. `Agents`, Tensors in the `Buffers` and random `tensors` in the `Trainers` (there should not be a lot of them though).

If you go through a few agents, you'll see that there's a bunch of sporadic `cpu` calls which would disrupt that. So I think thats the only other thing to be careful about that.

Let me know if you're planning to take it up or if you need more info about this.

**Comment by IshanDaga** [comment]
I'll raise a PR with some initial changes and we'll take it from there?

**Comment by Sharad24** [comment]
Sounds good!

**Comment by hades-rp2010** [comment]
Still working on this @IshanDaga ?

**Comment by IshanDaga** [comment]
Nope. Not at the moment

**Comment by hades-rp2010** [comment]
> Yes. Just ensure things that can be on the GPU, are on the GPU. E.g. `Agents`, Tensors in the `Buffers` and random `tensors` in the `Trainers` (there should not be a lot of them though).
> 
> If you go through a few agents, you'll see that there's a bunch of sporadic `cpu` calls which would disrupt that. So I think thats the only other thing to be careful about that.
> 
> Let me know if you're planning to take it up or if you need more info about this.

Would it be better to - 
1. Add a separate arg to the buffers, giving a choice to where to place them (ie GPU or CPU)
2. after sampling the list of tensors from the buffer, send these tensors to the GPU

I think 2 would be slower since we'd have to do a CPU-GPU transfer on every update step, but not sure if 1 is tractable as the buffer size grows larger
